lib/relation-definition.js

```diff
@@ -532,37 +532,85 @@ function lookupModel(models, modelName) {
+    params.as = params.as || modelToRef;  // modelToRef is relationTo name
+
     if (typeof modelTo === 'string') {
-      modelToName = (singularize ? i8n.singularize(params.as) : params.as).toLowerCase();
-      modelTo = lookupModel(modelFrom.dataSource.modelBuilder.models, modelToName) || modelTo;
+      // lookup modelTo by name
+      modelToName = modelTo;
+      modelToName = (singularize ? i8n.singularize(modelToName) : modelToName).toLowerCase();
+      modelTo = lookupModel(modelFrom.dataSource.modelBuilder.models, modelToName);
     }
-    if (typeof modelTo !== 'function') {
-      throw new Error(g.f('Could not find "%s" relation for %s', params.as, modelFrom.modelName));
+    if (!modelTo) {
+      // lookup modelTo by relationTo name
+      var relationToName = params.as;
+      modelToName = (singularize ? i8n.singularize(relationToName) : relationToName).toLowerCase();
+      modelTo = lookupModel(modelFrom.dataSource.modelBuilder.models, modelToName);
     }
+  } else {
+    // modelToRef might already be an instance of model
+    modelTo = modelToRef;
+  }
+  if (typeof modelTo !== 'function') {
+    throw new Error(g.f('Could not find relation %s for model %s', params.as, modelFrom.modelName));
   }
   return modelTo;
 }
 
-/*!
+/*
  * Normalize polymorphic parameters
- * @param {Object|String} params Name of the polymorphic relation or params
+ * @param {Object|String|Boolean} polymorphic Param `polymorphic` of the relation.
+ * @param {String} relationName The name of the relation we are currently setting up.
  * @returns {Object} The normalized parameters
  */
-function polymorphicParams(params, as) {
-  if (typeof params === 'string') params = {as: params};
-  if (typeof params.as !== 'string') params.as = as || 'reference'; // default
-  params.foreignKey = params.foreignKey || i8n.camelize(params.as + '_id', true);
-  params.discriminator = params.discriminator || i8n.camelize(params.as + '_type', true);
-  return params;
+function polymorphicParams(polymorphic, relationName) {
+  var reference;
+
+  if (typeof polymorphic === 'string') {
+    // relation type is different from belongsTo (hasMany, hasManyThrough, hasAndBelongsToMany, ...)
+    // polymorphic is the name of the matching belongsTo relation from modelTo to modelFrom
+    reference = polymorphic;
+  }
+
+  if (polymorphic === true) {
+    // relation type is belongsTo: the relation name is used as the polymorphic reference
+    reference = relationName;
+  }
+
+  // NOTE backward compatibility: keyword `as` was once used inside the `polymorphic` object
+  // this was removed to avoid confusion with keyword `as` at the relation definition root
+  if (polymorphic && typeof polymorphic.as === 'string') {
+    reference = polymorphic.as;
+  }
+
+  // relationName is eventually used as reference if provided and reference not already defined
+  // ultimately defaults to 'reference'
+  reference = reference || relationName || 'reference';
```

Comment:
back to the `polymorphic.as` keyword, prior to renaming.
the spec we agreed on mentions that the `as` should not be used when **building** belongsTo relation. That's true for end-user API.

But... internally, this keyword (now renamed to `reference`) is used for polymorphic through and polymorphic embeds relations. So we need to define this reference internally in the relation settings so those relations can work.

Just try removing it and some tests will start failing.